### PR TITLE
PARQUET-1692: [C++] LogicalType::FromThrift error on Centos 7 RPM

### PR DESCRIPTION
**Pull request by orozery**
See details here:
https://issues.apache.org/jira/browse/PARQUET-1692

### CODE REVIEW

**Comment by orozery** [comment]
Ran cmake-format, it made lots of changes (as you can see on the second commit).
Probably something messed up on my machine.
I'll appreciate anyone's help here (or if you can run cmake-format and push to my branch). Thanks!

**Comment by fsaintjacques** [comment]
You need a [specific](https://github.com/apache/arrow/blob/master/ci/docker/linux-apt-lint.dockerfile#L58) version (0.5.2) of cmake-format.

**Comment by orozery** [comment]
@fsaintjacques thanks!!

File: cpp/cmake_modules/FindThrift.cmake [hunk 1]
```diff
@@ -44,9 +44,9 @@ function(EXTRACT_THRIFT_VERSION)
       message(SEND_ERROR "Could not extract Thrift version. "
                          "Version output: ${THRIFT_VERSION}")
     endif()
-    set(THRIFT_VERSION "${CMAKE_MATCH_1}" PARENT_SCOPE)
+    set(FOUND_THRIFT_VERSION "${CMAKE_MATCH_1}" PARENT_SCOPE)
   else()
-    set(THRIFT_VERSION "${THRIFT_VERSION}" PARENT_SCOPE)
+    set(FOUND_THRIFT_VERSION "${THRIFT_VERSION}" PARENT_SCOPE)
```
Comment (kou) [inline cpp/cmake_modules/FindThrift.cmake]: We can use `THRIFT_VERSION` here.

File: cpp/cmake_modules/FindThrift.cmake [hunk 2]
```diff
@@ -93,17 +93,22 @@ find_package_handle_standard_args(Thrift
...
-                                   INTERFACE_INCLUDE_DIRECTORIES "${THRIFT_INCLUDE_DIR}")
-  if(WIN32 AND NOT MSVC)
-    # We don't need this for Visual C++ because Thrift uses
-    # "#pragma comment(lib, "Ws2_32.lib")" in
-    # thrift/windows/config.h for Visual C++.
-    set_target_properties(Thrift::thrift PROPERTIES INTERFACE_LINK_LIBRARIES "ws2_32")
-  endif()
-
   extract_thrift_version()
+  if(FOUND_THRIFT_VERSION VERSION_LESS ${CMAKE_FIND_PACKAGE_NAME}_FIND_VERSION)
```
Comment (kou) [inline cpp/cmake_modules/FindThrift.cmake]: We should not use our version check. We should use `find_package_handle_standard_args`'s `VERSION_VAR` feature: https://cmake.org/cmake/help/latest/module/FindPackageHandleStandardArgs.html

Example: https://github.com/apache/arrow/blob/master/cpp/cmake_modules/FindRapidJSONAlt.cmake#L59-L60

**Review by kou — CHANGES_REQUESTED** [review]
Thanks.
Could you use `find_package_handle_standard_args`'s `VERSION_VAR` instead?

And could you also remove Thrift related codes from RPM?

  * https://github.com/apache/arrow/blob/master/dev/tasks/linux-packages/apache-arrow/yum/arrow.spec.in#L47
  * https://github.com/apache/arrow/blob/master/dev/tasks/linux-packages/apache-arrow/yum/arrow.spec.in#L97-L99
  * https://github.com/apache/arrow/blob/master/dev/tasks/linux-packages/apache-arrow/yum/arrow.spec.in#L476-L478
  * https://github.com/apache/arrow/blob/master/dev/tasks/linux-packages/apache-arrow/yum/centos-7/Dockerfile#L54
  * https://github.com/apache/arrow/blob/master/dev/tasks/linux-packages/apache-arrow/yum/centos-8/Dockerfile#L61

**Comment by kou** [comment]
@xhochy @wesm Could you add Or Ozeri to contributors on JIRA to assign to this issue? I don't have permission to do this.

File: cpp/cmake_modules/FindThrift.cmake [hunk 3]
```diff
@@ -90,7 +90,9 @@ find_package_handle_standard_args(Thrift
                                   REQUIRED_VARS
                                   THRIFT_STATIC_LIB
                                   THRIFT_INCLUDE_DIR
-                                  THRIFT_COMPILER)
+                                  THRIFT_COMPILER
+                                  VERSION_VAR
+                                  ${CMAKE_FIND_PACKAGE_NAME}_FIND_VERSION)
```
Comment (kou) [inline cpp/cmake_modules/FindThrift.cmake]: `${CMAKE_FIND_PACKAGE_NAME}_FIND_VERSION)` is the version we specify by `resolve_dependency_with_version`. (`0.10.0` in this case.)
We need to specify version we detected here.

How about this?

```diff
diff --git a/cpp/cmake_modules/FindThrift.cmake b/cpp/cmake_modules/FindThrift.cmake
index a266ad215..f9d6296b4 100644
--- a/cpp/cmake_modules/FindThrift.cmake
+++ b/cpp/cmake_modules/FindThrift.cmake
@@ -62,6 +62,9 @@ if(Thrift_ROOT)
             PATHS ${Thrift_ROOT}
             PATH_SUFFIXES "include")
   find_program(THRIFT_COMPILER thrift PATHS ${Thrift_ROOT} PATH_SUFFIXES "bin")
+  if(THRIFT_COMPILER)
+    extract_thrift_version()
+  endif()
 else()
   # THRIFT-4760: The pkgconfig files are currently only installed when using autotools.
   # Starting with 0.13, they are also installed for the CMake-based installations of Thrift.
@@ -78,11 +81,15 @@ else()
                  HINTS ${THRIFT_PC_PREFIX}
                  NO_DEFAULT_PATH
                  PATH_SUFFIXES "bin")
+    set(THRIFT_VERSION ${THRIFT_PC_VERSION})
   else()
     find_library(THRIFT_STATIC_LIB thrift${THRIFT_MSVC_STATIC_LIB_SUFFIX}
                  PATH_SUFFIXES "lib/${CMAKE_LIBRARY_ARCHITECTURE}" "lib")
     find_path(THRIFT_INCLUDE_DIR thrift/Thrift.h PATH_SUFFIXES "include")
     find_program(THRIFT_COMPILER thrift PATH_SUFFIXES "bin")
+    if(THRIFT_COMPILER)
+      extract_thrift_version()
+    endif()
   endif()
 endif()
 
@@ -92,7 +99,7 @@ find_package_handle_standard_args(Thrift
                                   THRIFT_INCLUDE_DIR
                                   THRIFT_COMPILER
                                   VERSION_VAR
-                                  ${CMAKE_FIND_PACKAGE_NAME}_FIND_VERSION)
+                                  THRIFT_VERSION)
 
 if(Thrift_FOUND OR THRIFT_FOUND)
   set(Thrift_FOUND TRUE)
@@ -106,6 +113,4 @@ if(Thrift_FOUND OR THRIFT_FOUND)
     # thrift/windows/config.h for Visual C++.
     set_target_properties(Thrift::thrift PROPERTIES INTERFACE_LINK_LIBRARIES "ws2_32")
   endif()
-
-  extract_thrift_version()
 endif()
```

**Comment by wesm** [comment]
I assigned Or on JIRA

**Comment by orozery** [comment]
@kou thanks! I took your suggestion as-is.
I indeed did not propose a sound fix beforehand, as I misinterpreted the cmake documentation of find_package_handle_standard_args. 
Please double check on me since I really have no experience in cmake scripting.

**Comment by kou** [comment]
@ursabot crossbow submit -g linux

**Review by kou — APPROVED** [review]
+1

Thanks!
I'll merge this once Linux package builds are passed.

**Comment by kou** [comment]
@ursabot crossbow submit ubuntu-xenial

**Comment by kou** [comment]
All green. I'll merge this.

**Pull request closed without merging**